src/vs/editor/browser/services/treeSitter/treeSitterParserService.ts

```diff
@@ -201,40 +209,56 @@ export class TreeSitterTree implements IDisposable {
 
 export class TreeSitterLanguages extends Disposable {
 	private _languages: Map<string, Parser.Language> = new Map();
+	protected /*exposed for tests*/ readonly _onDidAddLanguage: Emitter<{ id: string; language: Parser.Language }> = this._register(new Emitter());
+	/**
+	 * If you're looking for a specific language, make sure to check if it already exists with `getLanguage` as it will kick off the process to add it if it doesn't exist.
+	 */
+	public readonly onDidAddLanguage: Event<{ id: string; language: Parser.Language }> = this._onDidAddLanguage.event;
 
 	constructor(private readonly _treeSitterImporter: TreeSitterImporter,
 		private readonly _fileService: IFileService,
-		private readonly _environmentService: IEnvironmentService
+		private readonly _environmentService: IEnvironmentService,
+		private readonly _registeredLanguages: Map<string, string>,
 	) {
 		super();
 	}
 
-	public async getLanguage(languageId: string): Promise<Parser.Language | undefined> {
+	public getLanguage(languageId: string): Parser.Language | undefined {
```

Comment:
I think it is surprising that `getLanguage` causes a side effect (triggering to load the language), but still returns `undefined`.
Can't you return `Promise<Parser.Language>` here? That would also simplify the call-site.

Reply:
I need it to be sync. I'll rename it.